In [54]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler

from imblearn.combine import SMOTEEN
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix, classification_report, plot_confusion_matrix

from sklearn.model_selection import train_test_split

import matplotlib.pyplot as plt
from sklearn.datasets import make_classification
from sklearn.model_selection import cross_validate

from sklearn import metrics
from sklearn.metrics import balanced_accuracy_score

### 1.	Perform combined over and undersampling on the diabetes dataset (use SMOTEENN). Explain how combined sampling works.

In [35]:
dia_df = pd.read_csv("../diabetes.csv")
dia_df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [49]:
X = dia_df.drop('Outcome', axis=1)
y = dia_df['Outcome']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=45,stratify=y)

sc= StandardScaler()

X_sc_train=sc.fit_transform(X_train)
X_sc_test=sc.fit_transform(X_test)


In [51]:
# Create the object
smote = SMOTEENN(random_state=45)
    
# Fit the data
X_resampled,y_resampled = smote.fit_resample(X_sc_train, y_train)


In [56]:
lr = LogisticRegression(random_state=45)

lr.fit(X_resampled,y_resampled)

y_pred = lr.predict(X_sc_test)

acc_score = balanced_accuracy_score(y_test,y_pred)

print("SMOTEENN performance score :",acc_score)

SMOTEENN performance score : 0.7801851851851851


### 2.	Comment on the performance of combined sampling vs the other approaches we have used for the diabetes dataset

#### As you can see performance of combined sampling using SMOTEENN method we get higher score whereas we get comparatively low performance score when we use regular logisitic regression on model. Score using SMOTEEEN is 78% vs 34% for logisitic regression only. So we can better performance when we use combined sampling method vs just logistic regression approach. 

In [39]:
lr = LogisticRegression()
sc= StandardScaler()

X_train=sc.fit_transform(X_train)
X_test=sc.fit_transform(X_test)

lr.fit(X_train,y_train)

y_predicted = lr.predict(X_test)

In [40]:
lr.score(X,y)

0.3489583333333333

In [41]:
print(classification_report(y_test, y_predicted))

              precision    recall  f1-score   support

           0       0.80      0.90      0.85       100
           1       0.76      0.59      0.67        54

    accuracy                           0.79       154
   macro avg       0.78      0.75      0.76       154
weighted avg       0.79      0.79      0.79       154



### 3.	What is outlier detection? Why is it useful? What methods can you use for outlier detection?

Outliers are extreme values that deviate from other observations on data, they may indicate a variability in a measurement, experimental errors or a novelty. Outlier detection is a process of identifying these extreme values from dataset. 

It is important to detect outliers because outliers may indicate bad data and they affect the mean and median which in turn affects the error in datatset. when you plot the data with outliers then you might get big deviations because of them. For specific algorithm like clustering, outliers can cause distortion of clusters.

There are methods to detect outliers but Standard deviation method and Interquartile range method are most common among them. If we know that the distribution of values in the sample is Gaussian or Gaussian-like, we can use the standard deviation of the sample as a cut-off for identifying outliers. The distribution of the data in this case is approximately normal. About 68% of the data values lie within one standard deviation of the mean and about 95% are within two standard deviations, and about 99.7% lie within three standard deviations. Therefore, if you have any data point that is more than 3 times the standard deviation, then those points are very likely to be anomalous or outliers.

Not all data is normal or normal enough to treat it as being drawn from a Gaussian distribution. In this case we can use interquartile range (IQR) method (visually boxplot) to detect outliers. The IQR is calculated as the difference between the 75th and the 25th percentiles of the data and defines the box in a box and whisker plot. The IQR can be used to identify outliers by defining limits on the sample values that are a factor k of the IQR below the 25th percentile or above the 75th percentile.

Other method includes DBscan clustering, Isolation forest. In DBscan clustering, it clusters data into groups.It is also used as a density-based anomaly detection method with either single or multi-dimensional data. Other clustering algorithms such as k-means and hierarchal clustering can also be used to detect outliers. Isolation forest is method for unsupervised learning to detect outliers.


### 4.	Perform a linear SVM to predict credit approval (last column) using this dataset: https://archive.ics.uci.edu/ml/datasets/Statlog+%28Australian+Credit+Approval%29 . Make sure you look at the accompanying document that describes the data in the dat file. You will need to either convert this data to another file type or import the dat file to python. You can use this code, but otherwise you follow standard practices we have already used many times: 

from sklearn.svm import SVC

classifier = SVC(kernel='linear')


In [42]:
aus_df = pd.read_csv("australian.csv")
aus_df

A1     A2      A3  A4  A5  A6     A7  A8  A9  A10  A11  A12  A13   A14  \
0     1  22.08  11.460   2   4   4  1.585   0   0    0    1    2  100  1213   
1     0  22.67   7.000   2   8   4  0.165   0   0    0    0    2  160     1   
2     0  29.58   1.750   1   4   4  1.250   0   0    0    1    2  280     1   
3     0  21.67  11.500   1   5   3  0.000   1   1   11    1    2    0     1   
4     1  20.17   8.170   2   6   4  1.960   1   1   14    0    2   60   159   
..   ..    ...     ...  ..  ..  ..    ...  ..  ..  ...  ...  ...  ...   ...   
685   1  31.57  10.500   2  14   4  6.500   1   0    0    0    2    0     1   
686   1  20.67   0.415   2   8   4  0.125   0   0    0    0    2    0    45   
687   0  18.83   9.540   2   6   4  0.085   1   0    0    0    2  100     1   
688   0  27.42  14.500   2  14   8  3.085   1   1    1    0    2  120    12   
689   1  41.00   0.040   2  10   4  0.040   0   1    1    0    1  560     1   

     A15  
0      0  
1      0  
2      0  
3      1  
4      1  
..   ...  
685    1  
686    0  
687    1  
688    1  
689    1  

[690 rows x 15 columns]

In [43]:
X = aus_df.drop('A15',axis=1)
y = aus_df['A15']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=45,stratify=y)

clf = SVC(kernel='linear')

svm_model = clf.fit(X_train,y_train)

y_pred = svm_model.predict(X_test)
print(y_pred)

[0 1 1 0 0 1 0 1 1 0 1 1 1 1 0 1 1 1 1 0 1 0 1 1 1 1 1 0 0 1 0 0 0 0 1 1 0
 1 1 0 1 0 1 1 1 1 1 1 1 1 1 1 1 0 1 1 0 1 0 0 1 0 0 1 1 1 0 0 0 0 1 0 1 1
 0 1 0 1 1 0 0 1 1 1 0 0 0 0 1 0 1 0 1 1 0 1 0 1 1 0 0 0 0 0 0 0 1 0 0 0 0
 1 0 0 0 1 1 0 1 1 0 0 1 0 0 1 1 0 1 0 0 0 0 1 0 0 1 1]


In [44]:
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[59 18]
 [ 7 54]]


In [45]:
print("Accuracy: ",metrics.accuracy_score(y_test,y_pred))

Accuracy:  0.8188405797101449


In [46]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.89      0.77      0.83        77
           1       0.75      0.89      0.81        61

    accuracy                           0.82       138
   macro avg       0.82      0.83      0.82       138
weighted avg       0.83      0.82      0.82       138



### 5.	How did the SVM model perform? Use a classification report. 

 SVM model perfomance was decent as you can see from below classification report. As you can see precision (accuracy of positive prediction) was 89 % for class 0 and 75 % for class 1. whereas recall (Fraction of positives that were correctly indentified) was higher on class 1 (89%) vs 77% for class 0. F1 score (what percent of positive predictions were correct) was close in both cases averaged about 82%. Support shows us total number of samples (y_test, which is 20% of total number of columns so 20% of 690 which is equal to 138) and in this report you can see that 77 out of 138 were class 0 and 61 out of 138 were class 1.

In [9]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.89      0.77      0.83        77
           1       0.75      0.89      0.81        61

    accuracy                           0.82       138
   macro avg       0.82      0.83      0.82       138
weighted avg       0.83      0.82      0.82       138



### 6.	What kinds of jobs in data are you most interested in? Do some research on what is out there. Write about your thoughts in under 400 words. 

There are so many types of roles in data science. Some of them are Data Scientist, Data Engineer, Data Analyst, Machine learning engineer, Business analyst and so on. Machine learning engineering interests me the most followed by data analyst and data engineer. 

Machine learning engineers make models to predict future outcomes. ML engineers are data- and metric-driven and have a strong foundation in mathematics and statistics. Machine learning engineers have to design and implement machine learning applications/algorithms such as clustering, anomaly detection, classification, or prediction to address business challenges. ML engineers build data pipelines, benchmark infrastructure, and do A/B testing.They work collaboratively with product and development teams to improve data quality via tooling, optimization, and testing. ML engineers have to monitor the performance and ensure the reliability of machine learning systems in the organization.

Data analysts are involved in data munging and data visualization. If there are requests from stakeholders, data analysts have to query databases. They are in charge of data that is scraped, assuring the quality and managing it. They have to interpret data and effectively communicate the findings. Data analysts are expected to know R, Python, HTML, SQL, C++, and Javascript. They need to be more than a little familiar with data retrieval and storing systems, data visualization and data warehousing using ETL tools, Hadoop-based analytics, and business intelligence concepts.

Data engineers ensure that an organization’s big data ecosystem is running without glitches for data scientists to carry out the analysis. Data infrastructure engineers develop, construct, test, and maintain highly scalable data management systems. Unlike data scientists who seek an exploratory and iterative path to arrive at a solution, data engineers look for the linear path. Data engineers will improve existing systems by integrating newer data management technologies. Data engineers collect and store data, do real-time or batch processing, and serve it for analysis to data scientists via an API. They log and handle errors, identify when to scale up, ensure seamless integration, and “build human-fault-tolerant pipelines.”